### ISSUE TITLE
Feedback for the Tech Challenge

### ISSUE THREAD

**Issue by marcoonroad** [jorgepd/desafio-dinheiro#1]
Hi, man! I'm from Stone and I'm in charge to give the proper feedback for your whole project. So, let's go:

---

You're failing in let us represent the existence of no money. Let's begin by your tests, in the case, `test/money_struct_test.exs`. In the lines 24 up to 34, we have:

```elixir
    #@tag :pending
    test "empty money" do
      assert Money.amount(%Money{}) ==
      {:ok, 0}
    end

    #@tag :pending
    test "zero" do
      assert Money.amount(%Money{amount: 0}) ==
      {:ok, 0}
    end
```

It's quite wrong. There's no coin worth 0 cents. There's an account with balance worth 0 cents, tho.
Also, you must not allow the representation of negative money. It's only related to accounts, for example, in the case of statement, debits, credits, whatever. It's only rules regarding the business, but money itself is not negative. There's no such thing as -200,00 R$ in my wallet. In the case, it's tested as well on lines 48-52:

```elixir
    #@tag :pending
    test "negative" do
      assert Money.amount(%Money{amount: -100}) ==
      {:ok, -100}
    end
```

To solve that, you could make such test cases returning an error, then breaking the test suite, and so
you could approach to solve that on file `lib/money_struct.ex` by replacing:

```elixir
defstruct amount: 0, code: "Bancor"
```

with:

```elixir
defstruct [:amount, :code]
```

Where all the fields are `nil` by default.

Later, in this same file (lines 65-67), you could use:

```elixir
  def money(amount, code) when
    is_number(amount) and
    amount > 0 and
    is_binary(code)
  do
    if code in ExRates.currency_codes do
      {:ok, %Money{amount: amount, code: code}}
    else
      {:error, "invalid currency code"}
    end
  end
```

(Where `ExRates.currency_codes/0` is the list of currency codes from `@ratio` constant.)

Instead of:

```elixir
  def money(amount, code) when is_number(amount) and is_binary(code) do
    {:ok, %Money{amount: amount, code: code}}
  end
```

---

Another problem is that you're using module names into the global namespace. Let's look the files:
- `lib/account_struct.ex` (module name just called `Account`)
- `lib/exchange_rates.ex` (module name just called `ExRates`)
- `lib/money_struct.ex` (module name just called `Money`)

Such module names could override any previous module defined and loaded by Elixir whenever the names match / are the same. Therefore, to avoid such global collisions, you could call them instead:
- `DesafioDinheiro.Account` rather than just `Account`
- `DesafioDinheiro.ExRates` rather than just `ExRates`
- `DesafioDinheiro.Money` rather than just `Money`

And thus, rewrite/refactor the due imports/aliases across your project's files.

---

The good parts are that you have separated the exchange rates module from the other modules. Therefore, it allow us to even perform some HTTP requests, save such returned up-to-date exchange rates and then retrieve them in-memory for later usage.

I've tried to link your documentation for module usage into ExUnit's `doctest`s, but I failed due some syntax/indentation errors. Nevertheless, it is not a requirement.

---

The huge/serious problem is that you're using **float division** to perform splits. The implementation
follows here (on `lib/desafio_dinheiro.ex`, lines 48-60):

```elixir
  defp do_split([head | tail], money, []) do
    with {:ok, amount} <- amount(money),
         {:ok, code} <- Money.code(money)
      do
      number = Enum.count(tail)
      new_amount = amount / number

      case money(new_amount, code) do
        {:ok, new_money} -> do_split(tail, new_money, [head])
        {:error, error} -> {:error, error}
      end
    end
end
```

We're gonna to likely decline any further projects which implement such bug/anti-feature. See:

- http://blog.plataformatec.com.br/2014/09/floating-point-and-currency/

Just see here how we implement division for such cases:

- https://github.com/liuggio/money/blob/master/lib/money.ex#L368
- https://github.com/kipcole9/money/blob/master/lib/money.ex#L749

---

Despite such problems, the project is quite good, there's some points which you could improve on, and I will reply any doubts here for sure.

Thanks man, and good luck! :+1:

**Comment by jorgepd** [comment]
I actually debated with myself on whether money should be only positive or not, and chose yes. It really makes sense what you said though. I am gonna fix that in order for money to be only positive, and account to be either positive, zero or negative.

Thanks for the feedback. I am gonna start working on all the problems you listed as soon as possible.

**Comment by marcoonroad** [comment]
There's no problem on how much time it takes. We're always here to follow your progress. :)

**Comment by jorgepd** [comment]
Hi,  @marcoonroad

I have just finished fixing the problems in this issue, thanks for highlighting them. But as I was reading one of your solutions for float division I encountered a problem I had when I first started this tech challenge.

My doubt is:
In Elixir specifically. Is it better to use functions to manipulate structs and guarantee its safety and modularity, in case one decides to redefine that struct later. As I did.
Or is it better to just use the struct itself, as in [here](https://github.com/liuggio/money/blob/master/lib/money.ex#L368).
The latter option makes the code looks a little more fluid and understandable, but it misses out on the points listed for the first one.
